testutil/isstrictnilchecker.go

```diff
@@ -0,0 +1,40 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutil
+
+import (
+	"reflect"
+
+	"gopkg.in/check.v1"
+)
+
+// IsStrictNil checks that the variable is nil typed and nil valued.
+var IsStrictNil = &isStrictNilChecker{
+	&check.CheckerInfo{Name: "IsStrictNil", Params: []string{"value"}},
+}
+
+type isStrictNilChecker struct {
+	*check.CheckerInfo
+}
+
+func (*isStrictNilChecker) Check(params []interface{}, names []string) (result bool, errMsg string) {
+	// both the type and value must be nil
+	return !reflect.ValueOf(params[0]).IsValid() && params[0] == nil, ""
```

Comment:
Could we return a message when `reflect.ValueOf(params[0]).IsValid()` and `reflect.ValueOf(params[0]).IsNil()`? To make it  clear that the problem is that it is the wrong nil rather than it is not nil at all. That could make it less puzzling for people debugging failures in tests.

Reply:
I omitted it because I thought the checker name could serve that purpose and it simplifies the code but I'll add something. Maybe something like `"variable type was not <nil>"` or maybe `expected <nil> type and value but got %T:%v` if that seems ok to you?